tensorflow/core/kernels/non_max_suppression_op.cc

```diff
@@ -195,6 +223,214 @@ void DoNonMaxSuppressionOp(
   std::copy_n(selected.begin(), selected.size(), output_indices_data.data());
 }
 
+void BatchedNonMaxSuppressionOp(
+    OpKernelContext* context, const Tensor& inp_boxes, const Tensor& inp_scores,
+    int num_boxes, const int max_size_per_class, const int total_size_per_batch,
+    const float score_threshold, const float iou_threshold,
+    bool pad_per_class = false) {
+  int q = inp_boxes.dim_size(2);
+  int num_classes = inp_scores.dim_size(2);
+  const int num_batches = inp_boxes.dim_size(0);
+
+  // Default clip window of [0, 0, 1, 1] if none specified
+  std::vector<float> clip_window{0, 0, 1, 1};
+
+  // [num_batches, per_batch_size * 4]
+  std::vector<std::vector<float>> nmsed_boxes(num_batches);
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<float>> nmsed_scores(num_batches);
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<float>> nmsed_classes(num_batches);
+  // [num_batches]
+  std::vector<int> final_valid_detections;
+
+  int per_batch_size = total_size_per_batch;
+
+  // perform non_max_suppression operation for each batch independently
+  for (int batch = 0; batch < num_batches; ++batch) {
+    // dims of per_batch_boxes [num_boxes, q, 4]
+    Tensor per_batch_boxes = inp_boxes.Slice(batch, batch + 1);
+    // dims of per_batch_scores [num_boxes, num_classes]
+    Tensor per_batch_scores = inp_scores.Slice(batch, batch + 1);
+
+    struct ResultCandidate {
+      int box_index;
+      float score;
+      int class_idx;
+      float box_coord[4];
+    };
+
+    std::vector<ResultCandidate> result_candidate_vec;
+
+    float* scores_data = per_batch_scores.unaligned_flat<float>().data();
+    float* boxes_data = per_batch_boxes.unaligned_flat<float>().data();
+
+    // Iterate through all classes
+    for (int class_idx = 0; class_idx < num_classes; ++class_idx) {
+      std::vector<float> class_scores_data;
+      std::vector<float> class_boxes_data;
+
+      for (int box = 0; box < num_boxes; ++box) {
```

Comment:
Before going in this for-loop make sure to 
1) ```class_scores_data.reserve(num_boxes);```
2) ```class_boxes_data.reserve(num_boxes*4);```

Reasoning:
1)Because we know the size, so prefer to reserve in-order to save on reallocation costs.
2)Similar to (1) but note, the way the code flows: it will always push ```num_boxes * 4``` elements into the vector ```class_boxes_data```.

Reply:
Thanks for the comment. I just pushed a commit that exactly does this.

Reply:
Perfect :)